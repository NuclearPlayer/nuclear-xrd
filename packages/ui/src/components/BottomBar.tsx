import { FC, ReactNode } from 'react';

interface BottomBarProps {
  children?: ReactNode;
  className?: string;
}

export const BottomBar: FC<BottomBarProps> = ({ children, className = '' }) => {
  return (
    <footer
      className={`h-16 bg-gray-900 border-t border-gray-700 flex items-center px-4 ${className}`}
    >
      {children}
    </footer>
  );
};
