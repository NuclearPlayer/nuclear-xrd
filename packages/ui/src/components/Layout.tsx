import { FC, ReactNode } from 'react';

interface PlayerShellProps {
  children: ReactNode;
  className?: string;
}

export const PlayerShell: FC<PlayerShellProps> = ({
  children,
  className = '',
}) => {
  return (
    <div
      className={`h-screen w-screen grid grid-rows-[auto_1fr_auto] ${className}`}
    >
      {children}
    </div>
  );
};
