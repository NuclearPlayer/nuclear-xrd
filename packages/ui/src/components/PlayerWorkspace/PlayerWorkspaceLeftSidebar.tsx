import { FC } from 'react';

import {
  PlayerWorkspaceSidebar,
  PlayerWorkspaceSidebarPropsBase,
} from './PlayerWorkspaceSidebar';

export const PlayerWorkspaceLeftSidebar: FC<
  PlayerWorkspaceSidebarPropsBase
> = ({ children, ...props }) => {
  return (
    <PlayerWorkspaceSidebar side="left" {...props}>
      {children}
    </PlayerWorkspaceSidebar>
  );
};
