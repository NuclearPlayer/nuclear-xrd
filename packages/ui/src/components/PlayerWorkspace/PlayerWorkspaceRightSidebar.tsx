import { FC } from 'react';

import {
  PlayerWorkspaceSidebar,
  PlayerWorkspaceSidebarPropsBase,
} from './PlayerWorkspaceSidebar';

export const PlayerWorkspaceRightSidebar: FC<
  PlayerWorkspaceSidebarPropsBase
> = ({ children, ...props }) => {
  return (
    <PlayerWorkspaceSidebar side="right" {...props}>
      {children}
    </PlayerWorkspaceSidebar>
  );
};
