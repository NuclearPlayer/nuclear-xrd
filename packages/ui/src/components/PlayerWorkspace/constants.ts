export const SIDEBAR_CONFIG = {
  COLLAPSED_WIDTH: 56,
  MIN_WIDTH: 150,
  MAX_WIDTH: 400,
  RESIZE_HANDLE_WIDTH: 1,
  TOGGLE_BUTTON_SIZE: 4,
  TOGGLE_BUTTON_OFFSET: 2,
} as const;

export const SIDEBAR_ANIMATIONS = {
  width: {
    spring: {
      type: 'spring' as const,
      stiffness: 300,
      damping: 30,
      mass: 0.8,
    },
    instant: { duration: 0 },
  },
  content: {
    spring: {
      type: 'spring' as const,
      stiffness: 400,
      damping: 25,
      mass: 0.5,
    },
    slideDistance: 20,
  },
  button: {
    hover: { scale: 1.1 },
    tap: { scale: 0.95 },
    spring: {
      type: 'spring' as const,
      stiffness: 400,
      damping: 17,
    },
    rotation: {
      type: 'spring' as const,
      stiffness: 300,
      damping: 20,
    },
  },
} as const;
