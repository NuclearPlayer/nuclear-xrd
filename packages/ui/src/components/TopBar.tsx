import { FC, ReactNode } from 'react';

type TopBarProps = {
  children?: ReactNode;
  className?: string;
};

export const TopBar: FC<TopBarProps> = ({ children, className = '' }) => {
  return (
    <header
      className={`h-12 bg-gray-900 border-b border-gray-700 flex items-center px-4 ${className}`}
    >
      {children}
    </header>
  );
};
