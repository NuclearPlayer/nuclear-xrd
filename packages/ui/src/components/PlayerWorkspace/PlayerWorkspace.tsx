import { FC, ReactNode } from 'react';

import { cn } from '../../utils';
import { PlayerWorkspaceLeftSidebar } from './PlayerWorkspaceLeftSidebar';
import { PlayerWorkspaceRightSidebar } from './PlayerWorkspaceRightSidebar';

type PlayerWorkspaceProps = {
  children: ReactNode;
  className?: string;
};

type MainProps = {
  children?: ReactNode;
  className?: string;
};

const PlayerWorkspaceMain: FC<MainProps> = ({ children, className = '' }) => {
  return (
    <main
      data-testid="player-workspace-main"
      className={cn('bg-background-secondary overflow-auto', className)}
    >
      {children}
    </main>
  );
};

type PlayerWorkspaceComponent = FC<PlayerWorkspaceProps> & {
  LeftSidebar: typeof PlayerWorkspaceLeftSidebar;
  RightSidebar: typeof PlayerWorkspaceRightSidebar;
  Main: typeof PlayerWorkspaceMain;
};

export const PlayerWorkspace: PlayerWorkspaceComponent = Object.assign(
  ({ children, className = '' }: PlayerWorkspaceProps) => {
    return (
      <div
        className={cn(
          'grid h-full min-h-0 grid-cols-[auto_1fr_auto] bg-background-secondary relative',
          className,
        )}
      >
        {children}
      </div>
    );
  },
  {
    LeftSidebar: PlayerWorkspaceLeftSidebar,
    RightSidebar: PlayerWorkspaceRightSidebar,
    Main: PlayerWorkspaceMain,
  },
);
