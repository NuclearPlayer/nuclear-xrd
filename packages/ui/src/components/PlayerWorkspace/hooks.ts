import { useCallback, useRef, useState } from 'react';

import { SIDEBAR_CONFIG } from './constants';

export const useSidebarResize = (
  width: number,
  onWidthChange: (width: number) => void,
  side: 'left' | 'right',
  isCollapsed: boolean,
) => {
  const isResizing = useRef(false);
  const [isResizingState, setIsResizingState] = useState(false);

  const handleMouseDown = useCallback(
    (e: React.MouseEvent) => {
      if (isCollapsed) return;

      isResizing.current = true;
      setIsResizingState(true);
      e.preventDefault();

      const startX = e.clientX;
      const startWidth = width;

      const handleMouseMove = (e: MouseEvent) => {
        if (!isResizing.current) return;

        const deltaX =
          side === 'left' ? e.clientX - startX : startX - e.clientX;
        const newWidth = Math.max(
          SIDEBAR_CONFIG.MIN_WIDTH,
          Math.min(SIDEBAR_CONFIG.MAX_WIDTH, startWidth + deltaX),
        );
        onWidthChange(newWidth);
      };

      const handleMouseUp = () => {
        isResizing.current = false;
        setIsResizingState(false);
        document.removeEventListener('mousemove', handleMouseMove);
        document.removeEventListener('mouseup', handleMouseUp);
      };

      document.addEventListener('mousemove', handleMouseMove);
      document.addEventListener('mouseup', handleMouseUp);
    },
    [width, onWidthChange, side, isCollapsed],
  );

  return { handleMouseDown, isResizingState };
};
