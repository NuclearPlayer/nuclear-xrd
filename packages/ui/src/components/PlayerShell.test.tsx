import { render } from '@testing-library/react';

import { BottomBar } from './BottomBar';
import { PlayerShell } from './Layout';
import { PlayerWorkspace } from './PlayerWorkspace';
import { TopBar } from './TopBar';

describe('PlayerShell', () => {
  it('matches snapshot with complete layout', () => {
    const { asFragment } = render(
      <PlayerShell>
        <TopBar />
        <PlayerWorkspace>
          <PlayerWorkspace.LeftSidebar
            width={200}
            isCollapsed={false}
            onWidthChange={() => {}}
            onToggle={() => {}}
          />
          <PlayerWorkspace.Main />
          <PlayerWorkspace.RightSidebar
            width={200}
            isCollapsed={false}
            onWidthChange={() => {}}
            onToggle={() => {}}
          />
        </PlayerWorkspace>
        <BottomBar />
      </PlayerShell>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
