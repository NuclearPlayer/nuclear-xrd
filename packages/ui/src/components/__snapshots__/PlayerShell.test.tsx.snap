// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerShell > matches snapshot with complete layout 1`] = `
<DocumentFragment>
  <div
    class="grid h-screen w-screen grid-rows-[auto_1fr_auto] overflow-hidden"
  >
    <header
      class="bg-background border-border flex h-12 items-center border-b-2 px-4"
    >
      <svg
        class="inline-flex h-6 w-6"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          stroke="currentColor"
          stroke-width="8"
          transform="matrix(.96157 0 0 .95989 1.9231 1.9198)"
        >
          <rect
            height="96"
            rx="19.2"
            ry="19.2"
            width="96"
            x="2"
            y="2"
          />
          <path
            d="m30 23.6 20 26.4 20-26.4zm20 26.4 20 26.4 20-26.4zm0 0h-40l20 26.4z"
          />
        </g>
      </svg>
      <div
        class="flex-1"
      />
      <button
        aria-checked="false"
        aria-label="Toggle theme"
        class="border-border relative inline-flex h-6 w-11 cursor-pointer items-center rounded-full border-2 transition-all focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-none data-checked:bg-primary bg-white data-disabled:cursor-not-allowed data-disabled:opacity-50 group"
        data-headlessui-state=""
        id="headlessui-switch-:r0:"
        role="switch"
        tabindex="0"
        type="button"
      >
        <span
          class="border-border pointer-events-none flex h-4 w-4 items-center justify-center rounded-full border-2 bg-white ring-0 transition-transform [&>svg]:fill-white dark:[&>svg]:fill-black translate-x-1 group-data-checked:translate-x-5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-sun text-foreground"
            fill="none"
            height="12"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="4"
            />
            <path
              d="M12 2v2"
            />
            <path
              d="M12 20v2"
            />
            <path
              d="m4.93 4.93 1.41 1.41"
            />
            <path
              d="m17.66 17.66 1.41 1.41"
            />
            <path
              d="M2 12h2"
            />
            <path
              d="M20 12h2"
            />
            <path
              d="m6.34 17.66-1.41 1.41"
            />
            <path
              d="m19.07 4.93-1.41 1.41"
            />
          </svg>
        </span>
      </button>
    </header>
    <div
      class="bg-background-secondary relative grid h-full min-h-0 grid-cols-[auto_1fr_auto]"
    >
      <div
        class="bg-background-secondary border-border relative flex flex-col p-2 border-r-2"
      >
        <span
          class="mb-4 flex flex-row items-center justify-end"
        >
          <button
            class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center top-2 px-2 right-1"
            data-headlessui-state=""
            data-testid="sidebar-toggle-left"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-panel-left"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M9 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          class="flex-1 overflow-hidden"
          style="opacity: 0; transform: translateX(-20px);"
        />
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors right-0"
        />
      </div>
      <main
        class="bg-background-secondary overflow-auto"
        data-testid="player-workspace-main"
      />
      <div
        class="bg-background-secondary border-border relative flex flex-col p-2 border-l-2"
      >
        <span
          class="mb-4 flex flex-row items-center justify-start"
        >
          <button
            class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center top-2 px-2 left-1"
            data-headlessui-state=""
            data-testid="sidebar-toggle-right"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-panel-right"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M15 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          class="flex-1 overflow-hidden"
          style="opacity: 0; transform: translateX(20px);"
        />
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors left-0"
        />
      </div>
    </div>
    <footer
      class="bg-background border-border flex h-16 items-center border-t-2 px-4"
    />
  </div>
</DocumentFragment>
`;
